✨ Mark specific KCP machines with delete annotation for scaling down

furkatgofurov7 opened the pull request:
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Adds a possibility to mark a specific control plane Machine to be deleted by a KCP when one or multiple machines are annotated by `delete-machine`  annotation.

**This PR:**
1. makes sure the oldest machine is deleted first when there is one or more machine is marked for deletion. It implements the prioritization logic in descending order with respect to rollout policies, while selecting machine for scale  down in the following order:
      - _outdatedMachines with the delete annotation_
      - _machines with the delete annotation_
      - _outdated machines_
      - _all machines_
2. adds a unit test.
3. updates the KCP proposal with a note describing the behavior for scale down with annotation.

Fixes #3910

furkatgofurov7 commented:
/cc @fabriziopandini

furkatgofurov7 commented:
/assign @fabriziopandini

vincepri commented:
/ok-to-test

vincepri commented on controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_types.go:
```diff
@@ -28,6 +28,10 @@ import (
 const (
 	KubeadmControlPlaneFinalizer = "kubeadm.controlplane.cluster.x-k8s.io"
 
+	// DeleteControlPlaneMachineAnnotation annotation deletes Control Plane machines with delete
+	// machine annotation key during the scaling down of KCP
+	DeleteControlPlaneMachineAnnotation = "kubeadm.controlplane.cluster.x-k8s.io/delete-machine"
```
Could we just reuse the `DeleteMachineAnnotation` defined in the core types?

vincepri replied:
I just realized that somehow that constant is defined in the `controllers` package. We should move it under the `api/v1alpha4` package instead, and then we should be able to reuse it here

fabriziopandini replied:
+1

furkatgofurov7 replied:
Thanks for the suggestion, I have made the changes accordingly.

vincepri commented:
/milestone v0.4.0

fabriziopandini commented on docs/proposals/20191017-kubeadm-based-control-plane.md:
```diff
@@ -336,7 +336,12 @@ spec:
   - See [Health checks](#Health checks) below.
-- Scale down operations removes the oldest machine in the failure domain that has the most control-plane machines on it 
+- Scale down operations removes the oldest machine in the failure domain that has the most control-plane machines on it
+- Scale down operations can be also triggered by marking specific control plane machine(s) using delete annotation key.
```
From what I understand, the annotation is a way to condition how to select which machine to delete when scaling down, but the annotation alone does not trigger scale down. 
If this is true, this sentence should be changed (or removed)

fabriziopandini commented on controlplane/kubeadm/internal/control_plane.go:
```diff
@@ -124,6 +124,14 @@ func (c *ControlPlane) MachineInFailureDomainWithMostMachines(machines Filterabl
 }
 
+// MachineWithDeleteAnnotation returns the machine that has delete annotatation key on it.
+func (c *ControlPlane) MachineWithDeleteAnnotation(machines FilterableMachineCollection) (*clusterv1.Machine, error) {
```
Can we avoid to return an error value, given that it is always nil?

fabriziopandini commented on controlplane/kubeadm/controllers/scale.go:
```diff
@@ -135,5 +135,9 @@ func selectMachineForScaleDown(controlPlane *internal.ControlPlane, outdatedMach
 	if outdatedMachines.Len() > 0 {
 		machines = outdatedMachines
 	}
+	withDeleteAnnotation, err := controlPlane.MachineWithDeleteAnnotation(machines)
+	if err == nil {
+		return withDeleteAnnotation, nil
+	}
```
```suggestion
	if withDeleteAnnotation, err := controlPlane.MachineWithDeleteAnnotation(machines); err == nil {
		return withDeleteAnnotation, nil
	}
```
Also given that error is always nil, should this check if there is at least one  machine withDeleteAnnotation?

furkatgofurov7 commented:
/retest

furkatgofurov7 commented:
@vincepri @fabriziopandini thanks, I have addressed all your comments(I accidentally amended my commit, sorry for that). However, tests are failing, and was wondering if some of them are just a flake in CI or I am missing something?

fabriziopandini commented:
@furkatgofurov7 
Failure in pull-cluster-api-apidiff-main is ok (it just report you moved the annotation).
Some test are not flakes
- lint errors in pull-cluster-api-make-main 
- following errors in pull-cluster-api-test-main 
```
--- FAIL: TestSelectMachineForScaleDown (0.00s)
    --- FAIL: TestSelectMachineForScaleDown/when_there_is_a_machine(s)_marked_with_incorrect_delete_annotation_key,_it_returns_only_machine(s)_with_proper_annotation (0.00s)
```

vincepri commented on controllers/machineset_delete_policy.go:
```diff
@@ -33,11 +33,8 @@ type (
 	// DeleteNodeAnnotation marks nodes that will be given priority for deletion
 	// when a machineset scales down. This annotation is given top priority on all delete policies.
-	// Deprecated: Please use DeleteMachineAnnotation instead.
+	// Deprecated: Please use DeleteMachineAnnotation instead under api/v1alpha4 package.
```
Revert this, after #3955 merges, we should rebase and this deprecated annotation is going to be removed from the codebase

vincepri commented on controlplane/kubeadm/controllers/scale.go:
```diff
@@ -135,5 +135,8 @@ func selectMachineForScaleDown(controlPlane *internal.ControlPlane, outdatedMach
 	if outdatedMachines.Len() > 0 {
 		machines = outdatedMachines
 	}
+	if withDeleteAnnotation := controlPlane.MachineWithDeleteAnnotation(machines); withDeleteAnnotation != nil {
+		return withDeleteAnnotation, nil
+	}
```
Let's move this at the top of this function, on line 134 given that the logic here effectively ignores any other criteria to pick a Machine to scale down, and always prefers the Machine that the user selected.

vincepri replied:
If we remove the .Oldest call inside `MachineWithDeleteAnnotation`, we can return the whole list from that function and overwrite the `machines` slice here. We should also move back this block after the `outdatedMachines` one, so it overwrites the list.

Then we can return normally

vincepri commented on controlplane/kubeadm/internal/control_plane.go:
```diff
@@ -124,6 +124,16 @@ func (c *ControlPlane) MachineInFailureDomainWithMostMachines(machines Filterabl
+	annotatedmachines := machines.Filter(machinefilters.HasAnnotationKey(clusterv1.DeleteMachineAnnotation))
+	// If there are, then pick the oldest machine from the filtered machines.
+	machineToPick := annotatedmachines.Oldest()
+
```
```suggestion
```

vincepri commented on controlplane/kubeadm/internal/control_plane.go:
```diff
@@ -124,6 +124,16 @@ func (c *ControlPlane) MachineInFailureDomainWithMostMachines(machines Filterabl
+func (c *ControlPlane) MachineWithDeleteAnnotation(machines FilterableMachineCollection) (*clusterv1.Machine) {
+	// See if there are any machines with DeleteMachineAnnotation key.
+	annotatedmachines := machines.Filter(machinefilters.HasAnnotationKey(clusterv1.DeleteMachineAnnotation))
+	// If there are, then pick the oldest machine from the filtered machines.
+	machineToPick := annotatedmachines.Oldest()
```
Hm, could we instead return the whole list of annotated machines here? And let the logic above pick the one in the failure domain with most machines?

furkatgofurov7 replied:
@vincepri we were discussing this in the issue itself and kind of agreed to return the oldest machine first. Are there any objections/drawbacks of don't it this way? would you mind elaborating more if so?  Thanks

vincepri replied:
`MachineInFailureDomainWithMostMachines` (what a mouthful 😂), picks the Machine that's in the failure domain (e.g. availability zone on AWS) with the highest count of machines in it, and then picks the oldest one from that list

furkatgofurov7 commented:
/test pull-cluster-api-test-main

furkatgofurov7 commented:
> @furkatgofurov7
> Failure in pull-cluster-api-apidiff-main is ok (it just report you moved the annotation).
> Some test are not flakes
> 
>     * lint errors in pull-cluster-api-make-main
> 
>     * following errors in pull-cluster-api-test-main
> 
> 
> ```
> --- FAIL: TestSelectMachineForScaleDown (0.00s)
>     --- FAIL: TestSelectMachineForScaleDown/when_there_is_a_machine(s)_marked_with_incorrect_delete_annotation_key,_it_returns_only_machine(s)_with_proper_annotation (0.00s)
> ```

@fabriziopandini @vincepri I have addressed all your comments, thanks for the review. Can you please suggest how can we fix pull-cluster-api-apidiff-main test?

fabriziopandini commented:
Failure in pull-cluster-api-apidiff-main is ok (it just report you moved the annotation); the job is not required, so it will not block this PR.

fabriziopandini commented:
/lgtm

detiber commented on controlplane/kubeadm/controllers/scale.go:
```diff
@@ -239,5 +239,8 @@ func selectMachineForScaleDown(controlPlane *internal.ControlPlane, outdatedMach
 	if outdatedMachines.Len() > 0 {
 		machines = outdatedMachines
 	}
+	if withDeleteAnnotation := controlPlane.MachineWithDeleteAnnotation(machines); withDeleteAnnotation.Len() > 0 {
+		machines = withDeleteAnnotation
+	}
```
Is the intention here to prioritize outdated Machines over Machines marked for deletion by annotation?

This would mean effectively that the annotation would be ignored if applied to a newer machine during the process of a rollout of changes until that rollout is complete.

I'm wondering if this will cause confusion for users.

I think it might make more sense to prioritize:
- any outdatedMachines with the delete annotation
- any machines with the delete annotation
- outdated machines
- all machines

```suggestion
	switch {
	case controlPlane.MachineWithDeleteAnnotation(outdatedMachines).Len() > 0:
		machines = controlPlane.MachineWithDeleteAnnotation(outdatedMachines)
	case controlPlane.MachineWithDeleteAnnotation(machines).Len() > 0:
		machines = controlPlane.MachineWithDeleteAnnotation(machines)
	case outdatedMachines.Len() > 0:
		machines = outdatedMachines
	}
```

vincepri replied:
That should be what the current code does, given that we replace the splice in each `if` block, but your switch case solution looks more readable to me 👍

detiber replied:
@vincepri there is a slightly different effect if the delete annotation is added to a machine that is up to date according to our rollout policies while there are any outdated machines that still exist.

The prioritization of the current code in the PR is:
- any outdatedMachines with the delete annotation
- outdated machines
- any machines with the delete annotation
- all machines

compared to my suggestion:
- any outdatedMachines with the delete annotation
- any machines with the delete annotation
- outdated machines
- all machines

vincepri replied:
Oh I see what you mean, yes that makes sense! ✅

furkatgofurov7 replied:
That makes sense to me as well, thanks for suggestion @detiber. I have addressed your comments. The original idea was giving priority to any machines(outdatedMachines or not) with annotation a top priority, as you have clarified above.  WDYT of this version? PTAL.

detiber commented on docs/proposals/20191017-kubeadm-based-control-plane.md:
```diff
@@ -336,7 +336,12 @@ spec:
-- Scale down operations removes the oldest machine in the failure domain that has the most control-plane machines on it 
+- Scale down operations removes the oldest machine in the failure domain that has the most control-plane machines on it
+- Allow scaling down of KCP with the possibility of marking specific control plane machine(s) to be deleted with annotation key.
+In order to delete a specific machine, user have to annotate a machine(s) that needs to be deleted and KCP implements the
+following:
+  - Filters out a machine(s) with machine delete annotation key.
+  - Makes sure to delete the oldest machine(s) first when more than one machine is marked for deletion.
```
This covers the simplistic case related to scale down, depending on the prioritization question I posed above it would also be good to include information how a rollout would be affected by the presence of the annotation.

detiber commented on controlplane/kubeadm/controllers/scale_test.go:
```diff
@@ -287,6 +296,20 @@ func TestSelectMachineForScaleDown(t *testing.T) {
 			expectErr:        false,
 			expectedMachine:  clusterv1.Machine{ObjectMeta: metav1.ObjectMeta{Name: "machine-3"}},
 		},
+		{
+			name:             "when there is a single machine marked with delete annotation key in machine collection, it returns only that marked machine",
+			cp:               annotatedControlPlane,
+			outDatedMachines: internal.NewFilterableMachineCollection(m6, m7),
+			expectErr:        false,
+			expectedMachine:  clusterv1.Machine{ObjectMeta: metav1.ObjectMeta{Name: "machine-7"}},
+		},
+		{
+			name:             "when there are multiple machines marked with delete annotation key in machine collection, it returns the oldest marked machine first",
+			cp:               annotatedControlPlane,
+			outDatedMachines: internal.NewFilterableMachineCollection(m7, m8),
+			expectErr:        false,
+			expectedMachine:  clusterv1.Machine{ObjectMeta: metav1.ObjectMeta{Name: "machine-8"}},
+		},
```
It would be good to also add test cases covering the expected behavior when an annotated machine is not in outdatedMachines but is part of annotatedControlPlane

detiber commented:
/hold
holding pending clarification around prioritization of the deletion annotation wrt rollout of changes.

detiber commented:
/lgtm
/hold cancel

@furkatgofurov7 thank you for the fixes, these changes lgtm

fabriziopandini commented:
/approve

furkatgofurov7 commented:
Thanks a lot, @detiber @fabriziopandini @vincepri for reviews/suggestions!
